### PR TITLE
[SPARK-12019][SPARKR] Support character vector for sparkR.init(), check param and fix doc

### PR DESCRIPTION
**Pull request by felixcheung**
and add tests.
Spark submit expects comma-separated list

### CODE REVIEW

**Comment by shivaram** [comment]
Hmm I think its better to just support both forms ? Isn't `c("jar1", "jar2")` more natural ?

**Comment by felixcheung** [comment]
It is. I could add that.

**Comment by sun-rui** [comment]
The example is correct. Jars is expected to a vector of character.

The bug is at  https://github.com/apache/spark/blob/master/R/pkg/R/client.R#L47.
The code can be change as follows:
```
  jars <- paste(jars, collapse = ",")
  if (jars != "") {
     jars <- paste0("--jars ", jars)
  }
```

It is more natural to use character vector in R. I don't think it necessary to support both forms as it complicates the logic.

**Comment by felixcheung** [comment]
I considered about changing it that way but thought that would break everyone who had gotten it to work with a comma-separated string.
Also spark-submit (pyspark, spark-shell, spark-submit, sparkR CLIs) is expecting a comma-separated string - it might make more sense for someone who is more familiar with that.

**Comment by felixcheung** [comment]
I could make it either way - let me know which way we should support.

**Comment by sun-rui** [comment]
My personal opinion is that comma-speparated jars is for command line options, while it is natural to use vector/array in API. jars is Seq[String] in Scala SparkContext API, for example.
If @shivaram insists both way, I am OK. It takes time to support them and more test cases are needed:)

**Comment by shivaram** [comment]
Well the only reason to support both use cases it to have some backwards compatibility for users who might already be using this. Can we see how much more complex the code becomes ? Its good to not break compatibility if we can avoid it.

**Comment by felixcheung** [comment]
it shouldn't be hard. `jars <- paste(jars, collapse = ",")` sun-rui suggested should gracefully handle the case length(jars) == 1 and has comma-separated values

**Comment by sun-rui** [comment]
@shivaram, from document's point of view, this is not backward compatibility issue:
```
@param sparkJars Character string vector of jar files to pass to the worker nodes.
```

The comma-separated string happens to work because there is a redudant code logic that pass the string to SparkSubmit --jars option, which I think is not necessary, as SparkContext.addJar will be called.

So I propose to remove the  redudant code logic.

**Comment by felixcheung** [comment]
Updated to support both "abc,def" and c("abc", "def") - we would need to remove empty string anyway (""), so not much more work to split by ','

**Comment by felixcheung** [comment]
@shivaram we might want this PR in Spark 1.6 ...

File: R/pkg/R/client.R [hunk 1]
```diff
@@ -44,12 +44,14 @@ determineSparkSubmitBin <- function() {
 }
 
 generateSparkSubmitArgs <- function(args, sparkHome, jars, sparkSubmitOpts, packages) {
+  jars <- paste0(jars, collapse = ",")
   if (jars != "") {
-    jars <- paste("--jars", jars)
+    jars <- paste0("--jars ", jars)
```
Comment (shivaram) [inline R/pkg/R/client.R]: This seems pretty subtle -- Can you add a comment just above saying `Leave a space after --jars`

File: R/pkg/R/sparkR.R [hunk 2]
```diff
@@ -366,3 +367,27 @@ getClientModeSparkSubmitOpts <- function(submitOps, sparkEnvirMap) {
   # --option must be before the application class "sparkr-shell" in submitOps
   paste0(paste0(envirToOps, collapse = ""), submitOps)
 }
+
+# Utility function to split by ',' and whitespace, remove empty tokens
+splitString <- function(input) {
```
Comment (shivaram) [inline R/pkg/R/sparkR.R]: Could we move this to utils.R ?

**Comment by shivaram** [comment]
Thanks @felixcheung -- This is very well done. I just had a couple of minor comments. You are right it will be good to have in 1.6 -- We'll merge this into branch-1.6 and if there is a RC2 we might catch 1.6.0. Lets see

cc @marmbrus

**Comment by shivaram** [comment]
LGTM. Merging this

**Pull request closed without merging**